src/Functions/mortonDecode.cpp

```diff
@@ -0,0 +1,427 @@
+            tuple_columns[2] = std::move(col2); \
+            tuple_columns[3] = std::move(col3); \
+            tuple_columns[4] = std::move(col4); \
+            tuple_columns[5] = std::move(col5); \
+            tuple_columns[6] = std::move(col6); \
+            return ColumnTuple::create(tuple_columns); \
+        case 8: \
+            tuple_columns[0] = std::move(col0); \
+            tuple_columns[1] = std::move(col1); \
+            tuple_columns[2] = std::move(col2); \
+            tuple_columns[3] = std::move(col3); \
+            tuple_columns[4] = std::move(col4); \
+            tuple_columns[5] = std::move(col5); \
+            tuple_columns[6] = std::move(col6); \
+            tuple_columns[7] = std::move(col7); \
+            return ColumnTuple::create(tuple_columns); \
+    } \
+    return ColumnTuple::create(tuple_columns);
+
+DECLARE_DEFAULT_CODE(
+constexpr auto MortonND_2D_Dec = mortonnd::MortonNDLutDecoder<2, 32, 8>();
+constexpr auto MortonND_3D_Dec = mortonnd::MortonNDLutDecoder<3, 21, 8>();
+constexpr auto MortonND_4D_Dec = mortonnd::MortonNDLutDecoder<4, 16, 8>();
+constexpr auto MortonND_5D_Dec = mortonnd::MortonNDLutDecoder<5, 12, 8>();
+constexpr auto MortonND_6D_Dec = mortonnd::MortonNDLutDecoder<6, 10, 8>();
+constexpr auto MortonND_7D_Dec = mortonnd::MortonNDLutDecoder<7, 9, 8>();
+constexpr auto MortonND_8D_Dec = mortonnd::MortonNDLutDecoder<8, 8, 8>();
+class FunctionMortonDecode : public IFunction
+{
+public:
+    static constexpr auto name = "mortonDecode";
+    static FunctionPtr create(ContextPtr)
+    {
+        return std::make_shared<FunctionMortonDecode>();
+    }
+
+    String getName() const override
+    {
+        return name;
+    }
+
+    size_t getNumberOfArguments() const override
+    {
+        return 2;
+    }
+
+    bool isSuitableForShortCircuitArgumentsExecution(const DataTypesWithConstInfo & /*arguments*/) const override { return false; }
+
+    ColumnNumbers getArgumentsThatAreAlwaysConstant() const override { return {0}; }
+
+    DataTypePtr getReturnTypeImpl(const ColumnsWithTypeAndName & arguments) const override
+    {
+        UInt64 tuple_size = 0;
+        const auto * col_const = typeid_cast<const ColumnConst *>(arguments[0].column.get());
+        if (col_const)
+        {
+            const auto * mask = typeid_cast<const ColumnTuple *>(col_const->getDataColumnPtr().get());
+            if (mask)
+            {
+                tuple_size = mask->tupleSize();
+                for (size_t i = 0; i < tuple_size; i++)
+                {
+                    auto ratio = mask->getColumn(i).getUInt(0);
+                    if (ratio > 8 || ratio < 1)
+                        throw Exception("Illegal argument " + arguments[0].column->getName()
+                                            + " of function " + getName()
+                                            + " should be >= 1 or <= 8",
+                                        ErrorCodes::ARGUMENT_OUT_OF_BOUND);
```

Comment:
Just a stylistic preference, but seems like something like this will better
```suggestion
                        throw Exception(ErrorCodes::ARGUMENT_OUT_OF_BOUND,
                            "Illegal argument {} of function {}, size of Tuple must be in range 1-8", 
                            arguments[0].column->getName(), getName());
```